src/v/cluster/tx_hash_ranges.h

```diff
@@ -0,0 +1,261 @@
+  model::partition_id tm_partition, int32_t partitions_amount) {
+    tx_hash_type range_size = get_default_range_size(partitions_amount);
+    tx_hash_type hash_range_begin = range_size * tm_partition;
+    tx_hash_type hash_range_end = (tm_partition + 1) == partitions_amount
+                                    ? tx_tm_hash_max
+                                    : (range_size * (tm_partition + 1)) - 1;
+    return {hash_range_begin, hash_range_end};
+}
+
+struct tx_hash_ranges_set
+  : serde::envelope<
+      tx_hash_ranges_set,
+      serde::version<0>,
+      serde::compat_version<0>> {
+    std::vector<tx_hash_range> ranges{};
+
+    tx_hash_ranges_set() = default;
+    tx_hash_ranges_set(std::vector<tx_hash_range>&& hr)
+      : ranges(hr) {}
+
+    auto serde_fields() { return std::tie(ranges); }
+
+    void add_range(tx_hash_range range) {
+        auto merged_to = ranges.end();
+        if (range.first != 0) {
+            for (auto r_it = ranges.begin(); r_it != ranges.end(); ++r_it) {
+                if (r_it->last == range.first - 1) {
+                    r_it->last = range.last;
+                    merged_to = r_it;
+                    break;
+                }
+            }
+        }
+        if (range.last != tx_tm_hash_max) {
+            for (auto r_it = ranges.begin(); r_it != ranges.end(); ++r_it) {
+                if (r_it->first == range.last + 1) {
+                    if (merged_to != ranges.end()) {
+                        merged_to->last = r_it->last;
+                        ranges.erase(r_it);
+                    } else {
+                        r_it->first = range.first;
+                        merged_to = r_it;
+                    }
+                    break;
+                }
+            }
+        }
+        if (merged_to == ranges.end()) {
+            ranges.push_back(range);
+        }
+    }
+
+    bool contains(tx_hash_type v) const {
+        for (const auto& r : ranges) {
+            if (r.contains(v)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool contains(const tx_hash_range& r) const {
+        for (const auto& range : ranges) {
+            if (range.contains(r)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool intersects(tx_hash_range range) const {
+        return std::any_of(
+          ranges.begin(), ranges.end(), [range1 = range](tx_hash_range range2) {
+              return range1.intersects(range2);
+          });
+    }
+};
+
+using repartitioning_id = named_type<int64_t, struct repartitioning_id_type>;
+
+namespace hosted_transactions {
+template<class T>
+tx_hash_ranges_errc
+exclude_transaction(T& self, const kafka::transactional_id& tx_id) {
+    if (self.excluded_transactions.contains(tx_id)) {
+        return tx_hash_ranges_errc::success;
+    }
+    if (self.included_transactions.contains(tx_id)) {
+        self.included_transactions.erase(tx_id);
```

Comment:
nit: eliminate an extra lookup by calling erase first, and examining the return value. similar patterns elsewhere.